Cache API enhancement

rajarshimaitra opened the pull request:
This adds convenience function to `Cache` to easily retrieve relevant asset information. This closes [Issue84](https://github.com/LNP-BP/rgb-node/issues/84).

Note:

- Asset listing function for cache is not separately implemented as there is already `assets()` API for `Cache` which does the same.

- `utxo_allocation_map(contract_id)` gives a map of `Map[Outpoint, Amount]` for a given asset identified by `contract_id`.

- `asset_allocation_map(utxo)` gives a map of `Map[Asset_name, Amount]` for a given `utxo`. There can be multiple `Asset` tied to a single `utxo`. Currently, the map only returns the asset name. This can be changed to any required field as needed.

- `asset_allocation_map(utxo)` is not implemented for `FileCache` as I didn't have enough time to play around with `FileCache`, this is marked as a `ToDo`.

- Test table for SQLite cache is updated to simulate many to many Asset-Utxo mapping. Test cases for the above two API are added.

dr-orlovsky commented on src/contracts/fungible/cache/cache.rs:
```diff
@@ -27,6 +29,18 @@ pub trait Cache {
+    fn utxo_allocation_map(
+        &self,
+        contract_id: ContractId,
+    ) -> Result<BTreeMap<bitcoin::OutPoint, AtomicValue>, Self::Error>;
```
Here we will replace the same outpoint if there were multiple assignments to it (see also comments below).

We should use `HashSet<bitcoin::OutPoint, Vec<AtomicValue>>` instead (`HashSet` since we do not need deterministic ordering and the size of the key space will be small)

dr-orlovsky commented on src/contracts/fungible/cache/cache.rs:
```diff
@@ -27,6 +29,18 @@ pub trait Cache {
+    fn asset_allocation_map(
+        &self,
+        utxo: &bitcoin::OutPoint,
+    ) -> Result<BTreeMap<String, u32>, CacheError>;
```
Same here as above, we need `HashSet` and `Vec<u32>` as values.

Regarding the keys, we need to index against ContractId, not String.

Also, why don't we use `AtomicAmount` here and use `u32`?

dr-orlovsky commented on src/contracts/fungible/cache/file.rs:
```diff
@@ -192,4 +192,34 @@ impl Cache for FileCache {
+                sum += alloc.value().value;
+            }
+
+            result.insert(item.0.clone(), sum);
```
This replaces any previous assignments to the same outpoint if they were existing, leading to incorrect return data. See my commend above about replacing the value with the `Vec`

dr-orlovsky commented on src/contracts/fungible/cache/file.rs:
```diff
@@ -192,4 +192,34 @@ impl Cache for FileCache {
+        &self,
+        _utxo: &bitcoin::OutPoint,
+    ) -> Result<BTreeMap<String, u32>, CacheError> {
+        unimplemented!();
```
This is quite simple: we have all data in memory in form of a data structure. We just need to access all assets data with `Cache::assets()` method and iterate over them collecting all items corresponding to the given outpoint.

dr-orlovsky commented on src/contracts/fungible/cache/sql.rs:
```diff
@@ -264,6 +264,82 @@ impl Cache for SqlCache {
+            for alloc in item.1 {
+                sum += alloc.value().value;
+            }
+            result.insert(item.0.clone(), sum);
```
Here we are also loosing previous value

dr-orlovsky commented on src/contracts/fungible/cache/sql.rs:
```diff
@@ -264,6 +264,82 @@ impl Cache for SqlCache {
+                .iter()
+                .fold(0, |acc, alloc| acc + alloc.amount as u32);
+
+            result.insert(asset_name, sum);
```
Again, loss of previous value

dr-orlovsky reviewed (changes requested):
Overall design looks good. Some specifics are addressed in comments; generally:
- we need to use vec of data for each asset type/output, instead of single value;
- use iterator over `assets()` returned data to collect all outpoint assignments with file cache

rajarshimaitra commented:
Yes, this was the basic framework, and I wanted to have your ideas regarding what and how data should be returned. Agreed with your comments with some added questions:

1. Did you mean `HashMap` [here](https://github.com/LNP-BP/rgb-node/pull/91#discussion_r512344232)? `HashSet` doesn't have key-value pairs.

2. What datatype should be returned as amount value? I have kept it as `u32` for now. Should we go for `AtomicAmount` or the full `Revealed` value which can be blinded later?

I will fix the summation issue.

dr-orlovsky commented:
1. Yeah, sorry, `HashMap` was meant
2. I think `AtomicAmount` (which is just a type alias for `u64`) will suffice. But `u32` is clearly not enough, since 32-bits is not sufficient to hold 64-bit amount.

dr-orlovsky commented:
Why type aliases for amounts are always preferable over native types is the fact that they denote unit of accounting, not only bit dimension. I.e. `AtomicAmount` is not native asset UoA, but the atomic UoA (like *satoshi* vs *bitcoin*)

rajarshimaitra commented:
Gotcha.. I have a slight feeling. All these `AtomicValue`, `AccountingValue`, `AccountingAmount` is getting a little ambiguous. I just found after searching that there is no `AtomicAmount`, only `AtomicValue`. For someone new looking at the code this might be a headache to figure out first. 

Is it possible to change the name of these variables to something more intuitive? Or we can simply have a doc patch somewhere explaining which one is what.

dr-orlovsky commented:
I agree this is hard to remember... However we need three types, and each of them must represent two dimensions: which unit of accounting it uses and which primitive type it uses to express amount in that unit of accounting. The types are:
* atomic unit ("satoshi"), always `u64`
* accounting unit ("bitcoin") using `f32` (easy to work in UI, but lossy)
* accounting unit ("bitcoin") using pair of `u64` for value and `u8` for precision (lossless, but hard to use in UI)

Feel free to propose best naming for them

rajarshimaitra commented:
Updated addressing all the comments.

The API now returns `HashMap`.

`asset_allocation_map()` implemented for `FileCache`.

Tests added for `FileCache` APIs. 

Running the tests is same as before:

- Create an env variable `DATABASE_URL`
- Comment  out `#[ignore]`s before tests.
- run `test/test_db.sh`

rajarshimaitra commented:
> Feel free to propose best naming for them

One way I can see to reduce this complexity is to treat `AccountingAmount` as an enum and have two variants describing two different ways of representing the same value. Then provide conversion between them. This would be more intuitive to understand that they are basically the same things. 
We also won't need two different variable names in this case, just two variants of `AccountingAmount` whose name can be suitably given.

dr-orlovsky commented on src/contracts/fungible/cache/cache.rs:
```diff
@@ -27,6 +29,18 @@ pub trait Cache {
     fn remove_asset(&mut self, id: ContractId) -> Result<bool, Self::Error>;
+
+    /// Returns the map of Utxo-Allocation_amount for a given asset
+    fn utxo_allocation_map(
```
```suggestion
    fn asset_allocations(
```

rajarshimaitra replied:
Fixed

dr-orlovsky commented on src/contracts/fungible/cache/cache.rs:
```diff
@@ -27,6 +29,18 @@ pub trait Cache {
+    ) -> Result<HashMap<bitcoin::OutPoint, Vec<AtomicValue>>, Self::Error>;
+
+    /// Returns the map of Asset-Allocation_amount for a given Outpoint
+    fn asset_allocation_map(
```
```suggestion
    fn output_assets(
```

rajarshimaitra replied:
Fixed

dr-orlovsky commented on src/contracts/fungible/cache/file.rs:
```diff
@@ -192,4 +192,178 @@ impl Cache for FileCache {
+        &self,
+        contract_id: ContractId,
+    ) -> Result<HashMap<bitcoin::OutPoint, Vec<AtomicValue>>, CacheError> {
+        let asset = self.asset(contract_id).unwrap();
```
Why we are unwrapping here?

rajarshimaitra replied:
Fixed

dr-orlovsky commented on src/contracts/fungible/cache/file.rs:
```diff
@@ -192,4 +192,178 @@ impl Cache for FileCache {
+    ) -> Result<HashMap<bitcoin::OutPoint, Vec<AtomicValue>>, CacheError> {
+        let asset = self.asset(contract_id).unwrap();
+
+        let allocation_map = asset.known_allocations();
+
+        let mut result = HashMap::new();
+
+        for item in allocation_map {
+            let mut alloc_amount = vec![];
+            for alloc in item.1 {
+                alloc_amount.push(alloc.value().value);
+            }
+            result.insert(item.0.clone(), alloc_amount);
+        }
```
```suggestion
        self
            .asset(contract_id)?
            .known_allocations()
            .into_iter()
            .map(|outpoint, allocations| allocations.into_iter().map(|a| a.value().value).collect())
            .collect();
```
is the same without unwraps, clones, unnamed tuple parts, new vecs and hash map allocations.

rajarshimaitra replied:
Fixed

dr-orlovsky commented on src/contracts/fungible/cache/file.rs:
```diff
@@ -192,4 +192,178 @@ impl Cache for FileCache {
+
+        for asset in self
+            .assets()
+            .unwrap()
```
unwrap!

rajarshimaitra replied:
fixed

dr-orlovsky commented on src/contracts/fungible/cache/file.rs:
```diff
@@ -192,4 +192,178 @@ impl Cache for FileCache {
+        for asset in self
+            .assets()
+            .unwrap()
+            .into_iter()
+            .cloned()
+            .collect::<Vec<Asset>>()
+        {
+            let allocations: Vec<AtomicValue> =
+                match asset.known_allocations().get(utxo) {
+                    Some(allocations) => allocations
+                        .into_iter()
+                        .map(|alloc| alloc.value().value)
+                        .collect(),
+                    None => continue,
+                };
+
+            result.insert(asset.id().clone(), allocations);
+        }
```
This also can be done without any single clone (btw ContractId is a copy type)

rajarshimaitra replied:
Ya its working without `clone()` but i need to at least deref the value because `asset.id()` gives a `&ContractId`

dr-orlovsky commented on src/contracts/fungible/cache/sql.rs:
```diff
@@ -264,6 +264,86 @@ impl Cache for SqlCache {
+        &self,
+        contract_id: ContractId,
+    ) -> Result<HashMap<bitcoin::OutPoint, Vec<AtomicValue>>, CacheError> {
+        // Get the required asset from cache
+        let asset = self.asset(contract_id)?;
+
+        // Fetch its known_allocation map
+        let allocation_map = asset.known_allocations();
+
+        // Process known_allocation map to produce the intended map
+        let mut result = HashMap::new();
+
+        for item in allocation_map {
+            let mut alloc_amount = vec![];
+            for alloc in item.1 {
+                alloc_amount.push(alloc.value().value);
+            }
+            result.insert(item.0.clone(), alloc_amount);
+        }
+
+        Ok(result)
```
Pls see sample from `FileCache`; the same applies here

rajarshimaitra replied:
fixed

dr-orlovsky commented on src/contracts/fungible/cache/sql.rs:
```diff
@@ -264,6 +264,86 @@ impl Cache for SqlCache {
+        let mut result = HashMap::new();
+
+        // Process the above groups to produce the required map
+        for item in utxo_allocation_groups {
+            let contract_id_string = sql_asset_table
+                .find(item.0.sql_asset_id)
+                .first::<SqlAsset>(&self.connection)
+                .map_err(|e| SqlCacheError::Sqlite(e))?
+                .contract_id
+                .clone();
+
+            let contract_id =
+                ContractId::from_hex(&contract_id_string[..]).unwrap();
+
+            let sum: Vec<AtomicValue> = item
+                .1
+                .iter()
+                .map(|alloc| alloc.amount as AtomicValue)
+                .collect();
+
+            result.insert(contract_id, sum);
+        }
```
It would be nice to use instead `item` just `(contract_id, amount)` and get rid of all those hard to read `.0` and `.1`s

rajarshimaitra replied:
fixed

dr-orlovsky reviewed (changes requested):
Enum will not work (a) because you will either return one type or another (b) and still need to reference them by using the some name, just after the `::`

rajarshimaitra commented:
Updated as per requested. Fixed unwraps that I missed. Better list comprehension for mapping logics added. 

 >Enum will not work (a) because you will either return one type or another (b) and still need to reference them by using the some name, just after the ::

Then I think the easiest thing we can do right now is to put [this](https://github.com/LNP-BP/rgb-node/pull/91#issuecomment-717352005) explanation as a doc in `data::assets` module.

dr-orlovsky reviewed (approved):
Thank you!
